virtualizarr/storage/_obstore.py

```diff
@@ -0,0 +1,246 @@
+from __future__ import annotations
+
+import pickle
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any
+
+from xarray import DataArray, Dataset
+from zarr.abc.store import (
+    ByteRequest,
+    OffsetByteRequest,
+    RangeByteRequest,
+    Store,
+    SuffixByteRequest,
+)
+from zarr.core.buffer import Buffer
+from zarr.core.buffer.core import BufferPrototype
+
+from virtualizarr.storage.common import (
+    find_matching_store,
+    get_zarr_metadata,
+    list_dir_from_xr_obj,
+    parse_manifest_index,
+)
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator, Iterable
+    from typing import Any
+
+    from obstore.store import ObjectStore as _UpstreamObjectStore
+    from zarr.core.buffer import BufferPrototype
+    from zarr.core.common import BytesLike
+
+
+__all__ = ["VirtualObjectStore"]
+
+_ALLOWED_EXCEPTIONS: tuple[type[Exception], ...] = (
+    FileNotFoundError,
+    IsADirectoryError,
+    NotADirectoryError,
+)
+
+
+class VirtualObjectStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
```

Comment:
The fact this uses `obstore` is an implementation detail, as it could for example later be changed to use `fsspec` with no user-facing changes (I'm not actually suggesting we do that). 

What makes this store unique is the fact is uses a manifest, which is why I wanted to call it `ManifestStore` (though `VirtualStore` is okay too).